### PR TITLE
docs: Add page about securing envoy to quick-start

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: docs: Add page about securing envoy to quick-start
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: .azure-pipelines/pipelines.yml [hunk 1]
```diff
@@ -59,7 +59,10 @@ stages:
               GCP_SERVICE_ACCOUNT_KEY: $(GcpServiceAccountKey)
             displayName: "Generate docs"
 
-          - script: ci/run_envoy_docker.sh 'ci/upload_gcs_artifact.sh /source/generated/docs docs'
+          - script: |
```
Comment (phlax) [inline .azure-pipelines/pipelines.yml:62]: remove workaround

**Comment by phlax** [comment]
@mattklein123 this one is nearly ready for final review - its taking a while as i was testing as i was writing it.

i still need to add/update some links and some small formatting/grammar issues

File: docs/root/start/quick-start/securing.rst [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+When connecting to an Envoy endpoint that is protected by ``SNI``, this must match one of the
+:ref:`server_names <envoy_v3_api_field_config.listener.v3.FilterChainMatch.server_names>` set in the endpoint's
+:ref:`filter_chain_match <envoy_v3_api_msg_config.listener.v3.FilterChainMatch>`, as
+:ref:`described above <start_quick_start_securing_sni>`.
+
+.. note::
+
+   When connecting to a ``TLS`` endpoint that is not protected by ``SNI``, this configuration is ignored, so it is
+   generally advisable to always set this to the DNS name of the endpoint you are connecting to.
```
Comment (phlax) [inline docs/root/start/quick-start/securing.rst:146]: not 100% sure if this is correct advice, but  seems to make sense to me

**Comment by phlax** [comment]
@mattklein123 this one should be ready for final review

File: docs/root/start/quick-start/securing.rst [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+Transport Layer Security (``TLS``) can be used to secure all types of ``HTTP`` traffic, including ``WebSockets``.
+
+Envoy also has support for transmitting and receiving generic ``TCP`` traffic over ``TLS``.
+
+.. warning::
+
+   The following guide takes you through individual aspects of securing traffic.
+
+   To secure traffic over a network that is untrusted, you are strongly advised to make
+   use of :ref:`SNI <start_quick_start_securing_sni>` *and* :ref:`mTLS <start_quick_start_securing_mtls>`
```
Comment (mattklein123) [inline docs/root/start/quick-start/securing.rst:18]: re: SNI, I would rephrase this, as I don't think it's required to use SNI. In some sense, SNI is just a routing concept (setting aside TLS1.3 encrypted SNI which breaks all of this).

re: mTLS, I might rephrase this slightly, as the way this is written will be contentious with some people. There is a set of people that think mTLS is too complicated and per-request (signing, etc. like JWT) is a better way to handle mutual authentication. I think mTLS is one way of doing this, but it would be good to rephrase slightly and potentially link to some of the per-request options such as JWT/Oauth/etc.

File: docs/root/start/quick-start/securing.rst [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+   this is what you should use when matching with ``match_subject_alt_names``.
+
+.. note::
+
+   See :ref:`here <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.CertificateValidationContext>` to view all
+   of the possible configurations for certificate validation.
+
+.. _start_quick_start_securing_sni:
+
+Secure an endpoint with SNI
```
Comment (mattklein123) [inline docs/root/start/quick-start/securing.rst:103]: Per above, SNI is about routing, not about security. It's actually not secure at all (until encrypted SNI in TLS 1.3). Can you check this around the doc and update/rephrase as needed?

File: docs/root/start/quick-start/securing.rst (hunk 2 above)
Comment (mattklein123) [reply docs/root/start/quick-start/securing.rst:146]: See above about SNI.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this is awesome. A few comments to get started.

/wait

File: docs/root/start/quick-start/securing.rst (hunk 3 above)
Comment (phlax) [reply docs/root/start/quick-start/securing.rst:18]: so re SNI, my understanding is that it prevents a type of spoofing - ie valid certs for wrong server (maybe this is wrong).

ill remove this/related advice and try to clarify what is written about it...

re mTLS - got it (i think) - ill make the warning about mutual authentication and encryption, and suggest this as a way to address those problems

File: docs/root/start/quick-start/securing.rst (hunk 4 above)
Comment (phlax) [reply docs/root/start/quick-start/securing.rst:103]: so i get its called routing but my understanding is that it just "addresses" the other side with an "sni" name

certainly in my testing, dns had no bearing here - as long as the sni name matched what was in the match list (or there was no  match list) then it worked

in terms of the title for this section - how about "Validating certificates with SNI" ?

i feel like we want to include sni here, and that it is sufficiently related to cert sec, but struggling to think how to word the section title

File: docs/root/start/quick-start/securing.rst (hunk 3 above)
Comment (phlax) [reply docs/root/start/quick-start/securing.rst:18]: > so re SNI, my understanding is that it prevents a type of spoofing - ie valid certs for wrong server (maybe this is wrong).

yep, from further testing, this is wrong.

sni "allows" multiple domains to be served at same ip with different certs. I guess browsers just automatically use the dns name here.

From envoy configuration pov i guess the advice is still that its best to put the dns of upstream here

Im less clear how important it is to include in sec section, but ill try and reword and we can figure what/whether to include

**Comment by phlax** [comment]
@mattklein123 ive updated the SNI sections, and the initial warning about using auth/encryption.

Im still not 100% about the sni stuff, but i think this is closer to what we want. 

I read the existing faq page, and anything else in the existing envoy docs. I tried playing with the `auto_sni` and `auto_san` settings, but couldnt seem to get them working or find any examples. The proto docs dont make it very clear the context that these can be used in.

I guess the main difference between what i have tried to add here and what is  in the faq already re sni, is why people would want/need to use it - from both server and client perspectives.

**Comment by phlax** [comment]
for reference these are the nginx docs related to sni https://docs.nginx.com/nginx/admin-guide/security-controls/terminating-ssl-http/#server-name-indication

**Comment by phlax** [comment]
also, not sure if this is necessary for first iteration - but would probs be good to get something about ocsp stapling in here

**Comment by phlax** [comment]
@mattklein123 it would be good to land this - ive posted it to users on slack twice in as many days - i think its pretty useful, and we can iterate

File: docs/root/start/quick-start/securing.rst [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+.. warning::
+
+   The following guide takes you through individual aspects of securing traffic.
+
+   To secure traffic over a network that is untrusted, you are strongly advised to make use of encryption
+   and mutual authentication wherever you control both sides of the connection or where relevant protocols are available.
+
+   Here we provide a guide to using :ref:`mTLS <start_quick_start_securing_mtls>` which provides both encryption
+   and mutual authentication.
```
Comment (mattklein123) [inline docs/root/start/quick-start/securing.rst:21]: We can do it as a follow up, but I think it would be good to cross link here to other authentication and authorization concepts such as JWT/RBAC/OAUTH/etc.

**Review by mattklein123 — APPROVED** [review]
Thanks this is great. Agreed we can ship and iterate. See my one follow up comment below. Thank you!

File: docs/root/start/quick-start/securing.rst (hunk 5 above)
Comment (phlax) [reply docs/root/start/quick-start/securing.rst:21]: cool, i have some other pending PRs around here, i reckon i can tack something on
